### ISSUE TITLE
The default behaviour of React component rendered under CKE shouldn't be changed

### ISSUE THREAD

**Issue by asimkt** [ckeditor/ckeditor5#8136]
I've added more details here: [Making focus works inside a CK Editor 5 createUIElement](https://stackoverflow.com/questions/63899538/making-focus-works-inside-a-ck-editor-5-createuielement)

When we allow users to render a React element (or any custom UI element for that matter) inside CKE, the expectation would be the component that rendered should behave as it was an outside component. But CK controls almost every event from the browser, this leads to different behaviour for the same component inside and outside CKE.

If CKE plan is to support React Component inside the editor as just a UI element without much functionality, can we mention the same in the [docs](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_downcastwriter-DowncastWriter.html#function-createUIElement) and the [react integration article](https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/frameworks/react.html)

### ✔️ Expected result

The react component behaviour shouldn't be controlled when it's inside CKE.

### ❌ Actual result

We cannot focus inside an input element if it's rendered inside CKE.

## 📃 Other details

* Browser: All
* OS: All
* CKEditor version: 20
* Installed CKEditor plugins: Essentials + RTC + some custom

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by FilipTokarski** [comment]
cc @pomek

**Comment by asimkt** [comment]
@pomek @FilipTokarski Found something:
https://github.com/ckeditor/ckeditor5-core/compare/proto/input-widget#diff-44ca1561ce575490eac0d660407d5144R239

Adding this to the plugin actually helps to do focus and other actions inside a widget. If anybody has more context on this that would be helpful. I'm happy to do a fix for this for all UIElement but I don't know whether this is the expected behaviour

**Comment by ma2ciek** [comment]
cc @jodator

**Comment by jodator** [comment]
@asimkt The thing you've found is a way to allow some events inside `UIElement` - it was researched to unblock editing form fields inside UIElement.

The issue should be solved by #4600 (the guide for [React in CKEditor 5 widget](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#demo) is also available). There's also a proposed solution (https://github.com/ckeditor/ckeditor5/issues/4600#issuecomment-648814894) that could work in your case without a need to modify CKEditor 5 code. However, I did not test that. And apparently `<select>` might not work in Firefox. If you'd only need input the linked code might be used in your converters.

As the work there is already done, maybe we could squeeze that to the next iteration. If you can check whether any of those proposed solutions would work with your case would be helpful. Additionally, could you share at least which events are you most interested in (besides all of them :wink: ) - we could use that to prototype test cases for most needed use-cases.

**Comment by jodator** [comment]
@asimkt The thing you've found is a way to allow some events inside `UIElement` - it was researched to unblock editing form fields inside UIElement.

The issue should be solved by #4600 (the guide for [React in CKEditor 5 widget](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#demo) is also available). There's also a proposed solution (https://github.com/ckeditor/ckeditor5/issues/4600#issuecomment-648814894) that could work in your case without a need to modify CKEditor 5 code. However, I did not test that. And apparently `<select>` might not work in Firefox. If you'd only need input the linked code might be used in your converters.

As the work there is already done, maybe we could squeeze that to the next iteration. If you can check whether any of those proposed solutions would work with your case would be helpful. Additionally, could you share at least which events are you most interested in (besides all of them :wink: ) - we could use that to prototype test cases for most needed use-cases.

**Comment by asimkt** [comment]
ok. I now know that I'm late to the party. 😄

The workaround from https://github.com/ckeditor/ckeditor5/issues/4465 does allow me to do focus on the input. (Tested)

Fix from [4600](https://github.com/ckeditor/ckeditor5/issues/4600#issuecomment-648814894), I need to test and I'll get back to you @jodator.

As far I can tell, these were the propositions:

- **Add a `data-ck-ignore-events` and handle it ourselves.**
For each action in the document, we will be running this check, and most of our actions will return false for the check which I don't think is that optimal (?)
- **Event listener for the element itself**
I think this is more performant as all other actions are not affected by this code.

And both are to be done at the per-user level. If we can have an option in `toWidget` like `options.ignoreCKEvents` and make this logic as part of CK itself, that will allow us:
- Progressively improve this solution. Including `<select/>` issue in FF.
- We'll have one source of truth to get all info related to this. Users may not need to update this fix for a newer browser/implementation issue. Fewer frictions on migrations also.

I myself see the integration of React inside CK is a powerful feature that has endless usages. But we need to let React takes care of all actions, and CK shouldn't have any control over that. For my use cases, I need:
1. Input boxes
2. A custom UI dropdown. (I think this uses select behind the scenes)
3. Full-screen popup on click.

But I think it's better to have a `completely ignore` event feature than having selectively removing individual events, (all of them 😄) because once the user starts integrating react components we may need to take care of all events anyway.


On a side note: can we link to #4600 from [React in CKEditor 5 widget page](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#demo)?

**Comment by jodator** [comment]
> But I think it's better to have a `completely ignore` event feature than having selectively removing individual events,

I agree here - I think that one of such improvements might be completly ignoring events coming from inside of `UIElement` and/or `RawElement`. We do "ignore" them in the editing view (like you can't put selection inside) so ignoring events coming from them would make sense.

And thanks for the summary! Hopefully, we could get back to this issue soon.

**Comment by asimkt** [comment]
thanks, @jodator Let me know if I can be of any help.

**Comment by Reinmar** [comment]
> Any plans to go with https://github.com/ckeditor/ckeditor5/compare/i/4600? @Reinmar

It's merged and released by now. 

> On a side note: can we link to #4600 from [React in CKEditor 5 widget page](https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#demo)? I spent some serious time looking for a bug on my side than an expected behaviour.

Done too: https://ckeditor.com/docs/ckeditor5/latest/framework/guides/tutorials/using-react-in-a-widget.html#before-you-start

As for whether we can make every UIElement/RawElement to treated as outside CKE5... I'm not entirely sure about this. Clicking a normal widget, even if it used a UIElement/RawElement to render its content (e.g. media embed) should still focus that widget. It only changes if that widget has UI inside. Maybe we could differentiate here UIElements and RawElements, but I feel that an attribute is just safer.

I'll close this ticket, but if you think that there's still a significant use case not covered, let us know by opening a new ticket.
